### PR TITLE
Condition route v3.1

### PR DESCRIPTION
**Pull request by wcy666103**
## What is the purpose of the change

Added multi-address and priority support

## Brief changelog

Detailed design documentation is available here  https://github.com/apache/dubbo/issues/14344

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/config/model/ConditionRuleParser.java [hunk 1]
```diff
@@ -40,14 +44,24 @@
...
+        if (confVersion != null && confVersion.startsWith(RULE_VERSION_V31)) {
+            rule = MultiDestConditionRouterRule.parseFromMap(map);
+            if (CollectionUtils.isEmpty(((MultiDestConditionRouterRule) rule).getConditions())) {
+                rule.setValid(false);
+            }
+        } else {
+            rule = ConditionRouterRule.parseFromMap(map);
+            if (CollectionUtils.isEmpty(((ConditionRouterRule) rule).getConditions())) {
+                rule.setValid(false);
+            }
```
Comment (chickenlj) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/config/model/ConditionRuleParser.java:77]: If rule version bigger than 3.1, should skip and print warning log.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/MultiDestConditionRouter.java [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        this.enabled = enabled;
+        matcherFactories =
+                moduleModel.getExtensionLoader(ConditionMatcherFactory.class).getActivateExtensions();
+        this.covert(multiDestCondition, this);
+        this.init(multiDestCondition.getFrom(), multiDestCondition.getTo());
+    }
+
+    public void init(Map<String, String> from, List<Map<String, String>> to) {
+        if (from == null || to == null) {
+            throw new IllegalArgumentException("Illegal route rule!");
```
Comment (chickenlj) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/MultiDestConditionRouter.java]: Make sure this exception will not affect RPC call execution. A warning log and skip should be fine.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/config/model/ConditionRuleParser.java (hunk 1 above)
Comment (wcy666103) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/config/model/ConditionRuleParser.java:77]: I have added that logic in the new version

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/MultiDestConditionRouter.java (hunk 2 above)
Comment (wcy666103) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/condition/MultiDestConditionRouter.java]: Here I keep the same logic as the old condition class(`ConditionStateRouter.class`), and I have the same try catch handling in the method caller ListenableStateRouter#process

**Review by chickenlj — APPROVED** [review]
LGTM.

**Comment by wcy666103** [comment]
My erroneous push resulted in several test commits, I will re-submit the pr here  https://github.com/apache/dubbo/pull/14356

**Pull request closed without merging**